src/self_driving/pilot_util.cpp

```diff
@@ -65,7 +199,6 @@ const std::list<metrics::PipelineMetricRawData::PipelineData> &PilotUtil::Collec
           common::ManagedPointer(accessor), exec_settings, metrics_manager);
 
       exec_ctx->SetParams(common::ManagedPointer<const std::vector<parser::ConstantValueExpression>>(&params));
-
       execution::compiler::ExecutableQuery::query_identifier.store(qid);
       auto exec_query = execution::compiler::CompilationContext::Compile(*out_plan, exec_settings, accessor.get(),
```

Comment:
Would it also work if we also reuse the `exec_ctx` and in the inner for loop only have only [`exec_ctx->SetParams`](https://github.com/cmu-db/noisepage/blob/5707742eb1264105c4d301c847248659517f46df/src/self_driving/pilot_util.cpp/#L201) and [`exec_query->Run`](https://github.com/cmu-db/noisepage/blob/5707742eb1264105c4d301c847248659517f46df/src/self_driving/pilot_util.cpp/#L205)?

I was using the assumption that we have pipeline data for the same query id nearby in pipeline metrics recorded and `pipeline_qids` recorded to adjust the qids [here](https://github.com/cmu-db/noisepage/blob/5707742eb1264105c4d301c847248659517f46df/src/self_driving/pilot_util.cpp/#L294-298). It was made quite implicit as a post-processing step since making a new copy of all pipeline metrics with the qids fixed directly here in [CollectPipelineFeatures](https://github.com/cmu-db/noisepage/blob/5707742eb1264105c4d301c847248659517f46df/src/self_driving/pilot_util.cpp/#L151) seem like too much effort.

Reply:
Got it. But I still think that if you execute multiple sets of parameters for the same query template without reusing the `ExecutableQuery`, this won't really work because you only record one qid [here](https://github.com/cmu-db/noisepage/blob/5707742eb1264105c4d301c847248659517f46df/src/self_driving/pilot_util.cpp/#L176).

I think reusing `ExecutableQuery` should solve this problem. And you do need to create the `exec_ctx` for each query execution because there's specific information for single query execution in the `exec_ctx`, such as some memory allocators.